### PR TITLE
adapter,compute,controller: attach workload_class label to relevant metrics

### PR DESCRIPTION
**Pull request by benesch**
This commit attaches a `workload_class` label whenever a metric is
associated with a workload with a known class (e.g., production,
staging). Relevant metrics are identified as followed:

  * All metrics in a `clusterd` process inherit the workload class of
    the cluster.

  * All metrics in an `environmentd` process that have an `instance_id`
    label inherit the workload class of the identified compute instance
    (cluster).

This will allow us to filter our monitoring based on workload class.

The implementation introduces a new API to the metrics registry called a
"postprocessor", which can mutate the state of all metrics on every call
to `MetricsRegistry::gather`. This was the most efficient implementation
available. In the future, we may want to consider a more explicit
approach of plumbing the `workload_class` label more directly to all
metrics that need it, but that would be an order of magnitude more work
than the approach taken here.

Part of https://github.com/MaterializeInc/materialize/issues/28097.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

This PR has no tests. That's plausibly fine, but not ideal. Let me know if you strongly object to me merging this PR without tests.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by teskje** [comment]
Things I've changed:
* In `ComputeController::dump`, stringify the `instance_workload_classes` entries to avoid serde stumbling over non-string keys.
* Make `ComputeController::update_instance_workload_class`, return an error rather than panicking when the instance is unknown. Also make it call `ComputeController::update_configuration` to remove a tiny bit of duplication.
* In `ComputeController::drop_instance`, remove the `instance_workload_classes` entry for the dropped instance.
* In `ComputeParameters`, make `workload_class` and `Option<Option<String>>`. The fields in this struct have the semantics of `None` meaning "parameter is unchanged", so we need a nested `Option` to also be able to communicate "parameter was unset".

**Comment by teskje** [comment]
My last commit also adds a test. Unfortunately, AFAICT the only way to get at clusterd metrics in a cluster test is using unmanaged clusters, for which we don't support for `ALTER CLUSTER`. I've added that support in the coordinator to make the test work, but I'm not sure about the implication so someone from @MaterializeInc/adapter should give their approval for this change before we merge it.

If the change is contentious we can also just drop the last commit and merge this PR without the test.

**Review by teskje — APPROVED** [review]

File: src/adapter/src/coord/sequencer/cluster.rs [hunk 1]
```diff
@@ -874,14 +875,27 @@ impl Coordinator {
...
+        let cluster = self.catalog().get_cluster(cluster_id);
+
+        let ops = vec![catalog::Op::UpdateClusterConfig {
+            id: cluster_id,
+            name: cluster.name().to_string(),
+            config: new_config,
+        }];
+
+        self.catalog_transact(Some(session), ops).await?;
+        Ok(())
```
Comment (benesch) [inline src/adapter/src/coord/sequencer/cluster.rs:898]: This change LGTM.

**Review by benesch — COMMENTED** [review]
Amazing, @teskje, thank you so much! The changes to support unmanaged clusters LGTM.

**Comment by benesch** [comment]
> If the change is contentious we can also just drop the last commit and merge this PR without the test.

I actually think we definitely want to support `ALTER CLUSTER ... SET (WORKLOAD CLASS = ...)` for unmanaged clusters, because there may be unmanaged clusters in production on which we need to set a workload class.

**Comment by teskje** [comment]
Nightly failures are all from https://github.com/MaterializeInc/materialize/issues/28519, afaict.

**Comment by benesch** [comment]
Yep, that was my assessment too. I think we’re good to merge!

**Pull request merged**